src/core/server/elasticsearch/cluster_client.ts

```diff
@@ -0,0 +1,186 @@
+
+  /**
+   * Indicates whether this cluster client (and all internal raw Elasticsearch JS clients) has been closed.
+   */
+  private isClosed = false;
+
+  constructor(private readonly config: ElasticsearchClientConfig, private readonly log: Logger) {
+    this.client = new Client(parseElasticsearchClientConfig(config, log));
+  }
+
+  /**
+   * Calls specified {@param endpoint} with provided {@param clientParams} on behalf of the
+   * Kibana internal user.
+   * @param endpoint String descriptor of the endpoint e.g. `cluster.getSettings` or `ping`.
+   * @param clientParams A dictionary of parameters that will be passed directly to the Elasticsearch JS client.
+   * @param options Options that affect the way we call the API and process the result.
+   */
+  public async callAsInternalUser(
+    endpoint: string,
+    clientParams: Record<string, unknown> = {},
+    options?: CallAPIOptions
+  ) {
+    this.assertIsNotClosed();
+
+    return await callAPI(this.client, endpoint, clientParams, options);
+  }
+
+  /**
+   * Closes the cluster client. After that client cannot be used and one should
+   * create a new client instance to be able to interact with Elasticsearch API.
+   */
+  public close() {
+    if (this.isClosed) {
+      return;
+    }
+
+    this.isClosed = true;
+    this.client.close();
+
+    if (this.scopedClient !== undefined) {
+      this.scopedClient.close();
+    }
+  }
+
+  /**
+   * Creates an instance of `ScopedClusterClient` based on the configuration the
+   * current cluster client that exposes additional `callAsCurrentUser` method
+   * scoped to the provided {@param req}. Consumers shouldn't worry about closing
+   * scoped client instances, these will be automatically closed as soon as the
+   * original cluster client isn't needed anymore and closed.
+   * @param req Request the `ScopedClusterClient` instance will be scoped to.
+   */
+  public asScoped(req: { headers?: Headers } = {}) {
+    // It'd have been quite expensive to create and configure client for every incoming
+    // request since it involves parsing of the config, reading of the SSL certificate and
+    // key files etc. Moreover scoped client needs two Elasticsearch JS clients at the same
+    // time: one to support `callAsInternalUser` and another one for `callAsCurrentUser`.
+    // To reduce that overhead we create one scoped client per cluster client and share it
+    // between all scoped client instances.
+    if (this.scopedClient === undefined) {
+      this.scopedClient = new Client(
+        parseElasticsearchClientConfig(this.config, this.log, {
+          auth: false,
+          ignoreCertAndKey: !this.config.ssl || !this.config.ssl.alwaysPresentCertificate,
+        })
+      );
+    }
+
+    const headers = req.headers
+      ? filterHeaders(req.headers, this.config.requestHeadersWhitelist)
+      : req.headers;
+
+    return new ScopedClusterClient(
+      (...args) => this.callAsInternalUser(...args),
```

Comment:
Hmm, I'm not sure what benefits `class properties` will bring comparing to good old arrow functions? I'm not opposed to it, but can you please elaborate on that a bit?

Reply:
- my very first question was why `callAsInternalUser` doesn't use this guard `assertIsNotClosed`? so I had to check it's implementation and see that okay, it's already there
- I had to go to `ScopedClusterClient` to check parameter names and understand that 2nd parameter it's just `scopedAPICaller`. Lambda doesn't pass intention in this case.
to recap - why we declare similar methods in different ways then?

Reply:
Okay, I see what you mean now. So you're proposing to have something like this, right?
```ts
  public callAsInternalUser = async (
    endpoint: string,
    clientParams: Record<string, unknown> = {},
    options?: CallAPIOptions
  ) => { ... };

  public asScoped(req: { headers?: Headers } = {}) {
    ......
    return new ScopedClusterClient(this.callAsInternalUser, this.callAsCurrentUser, headers);
  }

  private callAsCurrentUser = async (
    endpoint: string,
    clientParams: Record<string, unknown> = {},
    options?: CallAPIOptions
  ) => { ... };
```

If so, then yeah if it looks more readable to you, let's go this route. I'm just not very happy about `asScoped` implementation right now and believe it will change soon, the plan is to merge this PR as soon as possible (conflicting changes in upstream are appearing every other day) and then polish this part.